docs: improve documentation for mutable vs immutable Params (#3890)
Pyomo/pyomo: doc/OnlineDocs/explanation/modeling/math_programming/parameters.rst

   model.T = pyo.Param(model.A, validate=t_validate, initialize=t_data)

This example will produce the following error:

.. testoutput::

   Traceback (most recent call last):
     ...
   ValueError: Invalid parameter value: T[2] = '3', value type=<class 'int'>.
       Value failed parameter validation rule

Performance vs. Flexibility: Mutable Parameters
-----------------------------------------------

By default, Pyomo parameters are **immutable** (``mutable=False``). This choice 
is driven by performance:

* **Immutable (Default):** Pyomo "pre-computes" these values into the algebraic 
  expressions during model construction. This results in faster model generation 

>   and significantly lower memory usage, especially for large models.

Reviewer: There are a couple important things that "immutable" allows:
- We never create `ParamData` objects, which saves memory
- We directly inject the number into the expression tree, which we can detect and use to speed up walking expression trees
- Because the Param is "constant" and not just "fixed", the expression system is allowed to perform simplifications (e.g., if `p` and `q` are both immutable, then `5 * m.p * m.q[i]` will be simplified to a single float in the expression tree - which further speeds up model (expression) processing.

You hint at these.  I am not sure if there is value in expanding that exposition here -- or jif this is a detail better left to the section on Pyomo expressions?

* **Mutable:** Pyomo maintains the parameter as a symbolic object within 
  expressions. This allows you to change the value and re-solve without 
  rebuilding the entire model, but it adds computational overhead.

It is important to note that even immutable :class:`Param` objects carry some 
overhead. For the fastest possible model instantiation in a 
:class:`ConcreteModel`, using native Python data structures (like dictionaries 
or lists) to provide values directly into expressions is usually faster than 
using :class:`Param` components. However, as noted earlier, :class:`Param` 
provides benefits like validation and the ability to update values if declared 
as mutable.

When to use Mutable
~~~~~~~~~~~~~~~~~~~

**Use Immutable if:**

>   * The data is static and never changes during the execution of your script.

Reviewer: ```suggestion
  * The data is static and never changes during the lifetime of the model.
```

Author: done (Sorry I am doing these in such a way that you don't get credit for the commit, but you might have enough commits anyway :)

  * You want to maximize performance and minimize memory usage for large models.

**Use Mutable if:**
  * You are running a loop (e.g., sensitivity analysis) where you change 
    parameter values and re-solve.
  * You want to update values frequently without the "re-construction" 
    bottleneck.

>   * The parameter is part of a nonlinear expression that you need to update.

Reviewer: ```suggestion
  * The parameter is part of a nonlinear expression that you need to update.
  * You want named constants to be preserved in the Pyomo expressions (e.g., for documentation of debugging purposes)
```

Author: done.

Comparison: Param vs. Var
-------------------------

It is common to confuse mutable parameters with variables. The following table 
summarizes the key differences:

.. list-table::
   :header-rows: 1

   * - Feature
     - Param (Immutable)
     - Param (Mutable)

>      - Var

Reviewer: Would it be useful to split the `Var` column into `Var (fixed)` and `Var (free)`?  We sometimes get questions about "should I use a mutable Param or a fixed Var?"

Author: done (and an extremely good idea)

>    * - Can change after solve()?

Reviewer: ```suggestion
   * - Can change after model construction?
```

Author: done.
